<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Определение-и-заполнение-пропущенных-значений" data-toc-modified-id="Определение-и-заполнение-пропущенных-значений-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Определение и заполнение пропущенных значений</a></span></li><li><span><a href="#Преобразование-типов-данных" data-toc-modified-id="Преобразование-типов-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Преобразование типов данных</a></span></li><li><span><a href="#Поиск-дубликатов" data-toc-modified-id="Поиск-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Поиск дубликатов</a></span><ul class="toc-item"><li><span><a href="#Ручной-поиск-дубликатов-с-учетом-регистра" data-toc-modified-id="Ручной-поиск-дубликатов-с-учетом-регистра-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Ручной поиск дубликатов с учетом регистра</a></span></li><li><span><a href="#Ручной-поиск-и-удаление-явных-дубликатов" data-toc-modified-id="Ручной-поиск-и-удаление-явных-дубликатов-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Ручной поиск и удаление явных дубликатов</a></span></li><li><span><a href="#Проверка-столбцов-с-категориальными-переменными" data-toc-modified-id="Проверка-столбцов-с-категориальными-переменными-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Проверка столбцов с категориальными переменными</a></span></li><li><span><a href="#Лемматизация-данных" data-toc-modified-id="Лемматизация-данных-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>Лемматизация данных</a></span></li><li><span><a href="#Категоризация-столбца-'purpose'" data-toc-modified-id="Категоризация-столбца-'purpose'-2.3.5"><span class="toc-item-num">2.3.5&nbsp;&nbsp;</span>Категоризация столбца 'purpose'</a></span></li></ul></li><li><span><a href="#Создание-словарей" data-toc-modified-id="Создание-словарей-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Создание словарей</a></span></li></ul></li><li><span><a href="#Проверка-гипотез" data-toc-modified-id="Проверка-гипотез-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка гипотез</a></span><ul class="toc-item"><li><span><a href="#Проверка-зависимости-между-наличием-детей-и-возвратом-кредита-в-срок" data-toc-modified-id="Проверка-зависимости-между-наличием-детей-и-возвратом-кредита-в-срок-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Проверка зависимости между наличием детей и возвратом кредита в срок</a></span></li><li><span><a href="#Проверка-зависимости-между-семейным-положением-и-возвратом-кредита-в-срок" data-toc-modified-id="Проверка-зависимости-между-семейным-положением-и-возвратом-кредита-в-срок-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка зависимости между семейным положением и возвратом кредита в срок</a></span></li><li><span><a href="#Проверка-зависимости-между-уровнем-дохода-и-возратом-кредита-в-срок" data-toc-modified-id="Проверка-зависимости-между-уровнем-дохода-и-возратом-кредита-в-срок-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Проверка зависимости между уровнем дохода и возратом кредита в срок</a></span></li><li><span><a href="#Проверка-завимости-целей-кредита-влияют-на-возврат-в-срок" data-toc-modified-id="Проверка-завимости-целей-кредита-влияют-на-возврат-в-срок-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Проверка завимости целей кредита влияют на возврат в срок</a></span></li></ul></li><li><span><a href="#Итоги-исследования" data-toc-modified-id="Итоги-исследования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Итоги исследования</a></span></li></ul></div>

# Задачи проекта

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платежеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Обзор данных

Первоначально импортируем необходимую библиотеки и сохраним в переменную таблицу.

In [1]:
# импортируем библиотеки
import pandas as pd
from pymystem3 import Mystem

In [2]:
# чтение файла с данными с сохранение в переменную 'df'
try:
    df = pd.read_csv('/datasets/data.csv') 
except:
    df = pd.read_csv('D:\datasets\project2_df.csv')
    
df.head(10) # посмотрим

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Применим метод info(), чтобы посмотреть общую информацию о таблице.

In [3]:
df.info() # посмотрим

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице 12 столбцов, все они названы согласно змеиному_регистру. Типы данных в столбцах: **int64**, **float64**, **object**. 

**Согласно документации к данным:**
- `children` — количество детей в семье;
- `days_employed` — общий трудовой стаж в днях;
- `dob_years` — возраст клиента в годах;
- `education` — уровень образования клиента;
- `education_id` — идентификатор уровня образования;
- `family_status` — семейное положение;
- `family_status_id` — идентификатор семейного положения;
- `gender` — пол клиента;
- `income_type` — тип занятости;
- `debt` — имел ли задолженность по возврату кредитов;
- `total_income` — ежемесячный доход;
- `purpose` — цель получения кредита.

**Наблюдения:**
- в столбце `days_employed` встречается много отрицательных значений;
- в столбце `education` разный стиль заполнения;
- в значениях столбца `total_income` фигурируют 6 знаков после запятой;
- в столбцах `days_employed` и `total_income` есть пропущенные значения;
- строка с индексом 4: столбец `days_employed` трудовой стаж в днях - большая величина значений.

## Предобработка данных

### Определение и заполнение пропущенных значений

Посмотрим количество пропущенных значений.

In [4]:
df.isna().sum() # посмотрим

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**2174** пропущенных значений в столбцах **days_employed** и **total_income**. В пятой строке столбца **days_employed** значение равно **340266.072047** дням, что невозможно! Необходимо проверить количество подобных значений в этом столбце.

In [5]:
# посмотрим на строки с пропущенными значениями
df[df.isna().any(1)].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Данные в столбцах **days_employed** и **total_income** отсутствуют одновременно. Переменные в этих столбцах - количественные, для заполнения пропущенных значений подойдут среднее арифметическое или медиана.

Предположим, если работать с **18** до **80** лет, максимальный рабочий стаж составляет **62** года или **22072** дня. Посчитаем количество строк с значениями, превышающими данный показатель.

In [6]:
# уберем отрицательные значений в столбце 'days_employed'
df['days_employed'] = df['days_employed'].abs()

# посчитаем количество значений
print('Количество значений в столбце "days_employed" больше 22072:',(df[df['days_employed'] > 22072]['days_employed'].count())) 

df[df['days_employed'] > 22072].head(10) # посмотрим

Количество значений в столбце "days_employed" больше 22072: 3445


children  days_employed  dob_years education  education_id  \
4          0  340266.072047         53   среднее             1   
18         0  400281.136913         53   среднее             1   
24         1  338551.952911         57   среднее             1   
25         0  363548.489348         67   среднее             1   
30         1  335581.668515         62   среднее             1   
35         0  394021.072184         68   среднее             1   
50         0  353731.432338         63   среднее             1   
56         0  370145.087237         64   среднее             1   
71         0  338113.529892         62   среднее             1   
78         0  359722.945074         61    высшее             0   

            family_status  family_status_id gender income_type  debt  \
4        гражданский брак                 1      F   пенсионер     0   
18         вдовец / вдова                 2      F   пенсионер     0   
24  Не женат / не замужем                 4      F   пенсионер     0   
25        женат / замужем                 0      M   пенсионер     0   
30        женат / замужем                 0      F   пенсионер     0   
35       гражданский брак                 1      M   пенсионер     0   
50        женат / замужем                 0      F   пенсионер     0   
56         вдовец / вдова                 2      F   пенсионер     0   
71        женат / замужем                 0      F   пенсионер     0   
78        женат / замужем                 0      M   пенсионер     0   

     total_income                                purpose  
4   158616.077870                        сыграть свадьбу  
18   56823.777243     на покупку подержанного автомобиля  
24  290547.235997  операции с коммерческой недвижимостью  
25   55112.757732                   покупка недвижимости  
30  171456.067993  операции с коммерческой недвижимостью  
35   77805.677436                  на проведение свадьбы  
50   92342.730612                             автомобили  
56  149141.043533                            образование  
71   43929.696397                             автомобили  
78  175127.646000                   сделка с автомобилем

Посмотрим уникальные значения столбца **income_type** в таблице, в которой значения столбца **days_employed** больше **22072**.

In [7]:
display(df[df['days_employed'] > 22072]['income_type'].unique()) # посмотрим

array(['пенсионер', 'безработный'], dtype=object)

Все аномально высокие показатели столбца **days_employed** происходят при значении столбца **income_type** **пенсионер** или **безработный**. Из-за чего я могу предположить, что существует ошибка заполнения данных, если статус человека **неработающий**.

В столбце **days_employed** **2174** значений пропущены и **3445** значений с некорректным содержанием, итого: **5619** из **21525** строк датафрема (**26%**). Заполненение средним арифметическим или медианой считаю некорректным в таком искажении данных. 

Так как столбец **days_employed** не понадобится для проверки гипотез поставленных в задаче, считаю правильным заменить все **NaN** на ноль.

In [8]:
# заменяем все 'NaN' на ноль в 'столбце days_employed'
df['days_employed'] = df['days_employed'].fillna(0)

# считаем среднее арифметическое и медиану значений столбца 'total_income'
total_income_mean = df['total_income'].mean()
total_income_median = df['total_income'].median()

print('Cреднее арифметическое значение:', total_income_mean)
print('Медианное значение:', total_income_median) 
print()
# выведем на экран минимальное и максимальное значения столбца total_income
print('Минимальное значение столбца "total_income":', df['total_income'].min())
print('Максимальное значение столбца "total_income":', df['total_income'].max())

Cреднее арифметическое значение: 167422.30220817294
Медианное значение: 145017.93753253992

Минимальное значение столбца "total_income": 20667.26379327158
Максимальное значение столбца "total_income": 2265604.028722744


Выведем первые 10 самых больших значений столбца **total_income**.

In [9]:
df['total_income'].sort_values(ascending=False).head(10) # посмотрим

12412    2.265604e+06
19606    2.200852e+06
9169     1.726276e+06
20809    1.715018e+06
17178    1.711309e+06
17503    1.597613e+06
18368    1.551153e+06
18353    1.427934e+06
15268    1.350246e+06
11071    1.286281e+06
Name: total_income, dtype: float64

Из-за сильного перепада максимальных значений столбца **total_income** более корректно будет будет использовать медианное значение для заполнения пропусков в данном столбце.

In [10]:
# проверим  и посчитаем уникальные значения столбца 'income_type'
df['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

Мы видим, что в таблице подавляющее большинство значений столбца **income_type**: **сотрудник**, **компаньон**, **пенсионер**, **госслужащий**. Значения **предприниматель**, **безработный**, **в декрете**, **студент** в сумме встречаются всего **6** раз.

In [11]:
# преобразуем значения в столбце 'education'
df['education'] = df['education'].str.lower()

# сгруппируем по столбцу 'education' и посмотрим количество значений столбца 'income_type'
df.groupby('education')['income_type'].value_counts()

education            income_type    
высшее               сотрудник          2341
                     компаньон          1788
                     пенсионер           577
                     госслужащий         550
                     предприниматель       2
                     безработный           1
                     студент               1
начальное            сотрудник           145
                     пенсионер            99
                     компаньон            30
                     госслужащий           8
неоконченное высшее  сотрудник           370
                     компаньон           281
                     госслужащий          52
                     пенсионер            41
среднее              сотрудник          8260
                     пенсионер          3137
                     компаньон          2986
                     госслужащий         848
                     безработный           1
                     в декрете             1
ученая степень    

Для подсчета медианного значения возьмем комбинацию параметров **образование** и **профессия**. Создадим функцию, в которой заменим пропущенные значения столбца **total_income** данной комбинацией.

In [12]:
def all_at_once(education):
    for i in df['income_type'].unique():
        df.loc[(df['income_type'] == i)&(df['education'] == education)&(df['total_income'].isna()), 'total_income'] = \
        df.loc[(df['income_type'] == i)&(df['education'] == education), 'total_income'].median()

# создадим цикл, который будет проходить по уникальным значениям 'education' и запускать функцию 'all_at_once'
for k in df['education'].unique():
    all_at_once(k)

df['total_income'].isna().sum() # проверим

0

### Преобразование типов данных

Изменим тип данных в столбцах **days_employed** и **total_income** на **int64**.

In [13]:
df['days_employed'] = df['days_employed'].astype('int64')
df['total_income'] = df['total_income'].astype('int64') # преобразуем 

df[['days_employed', 'total_income']].dtypes # проверим

days_employed    int64
total_income     int64
dtype: object

### Поиск дубликатов

#### Ручной поиск дубликатов с учетом регистра

Создадим функцию **register**, которая будет считать уникальные значения конкретного столбца.

In [14]:
def register(column):
    return df[column].value_counts()

register('family_status') # проверим значения столбца 'family_status'

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [15]:
register('income_type') # проверим значения столбца 'income_type'

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

In [16]:
register('purpose') # проверим значения столбца 'purpose'

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
жилье                                     647
покупка жилья                             647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

На основании работы функции мы обнаружили, что значения в столбце **family_status** необходимо привести к нижнему регистру. С столбцом **purpose** необходимо произвести лемматизацию. В столбце **income_type** значения не пересекаются и соответствуют нижнему регистру.

In [17]:
df['family_status'] = df['family_status'].str.lower() # преобразуем

#### Ручной поиск и удаление явных дубликатов

Проверим данные на явные дубликаты.

In [18]:
print('Дубликатов в таблице:', df.duplicated().sum()) # проверим

Дубликатов в таблице: 71


Необходимо удалить явные дубликаты.

In [19]:
df = df.drop_duplicates().reset_index(drop=True) # удалим

print('Дубликатов в таблице:', df.duplicated().sum()) # проверим

Дубликатов в таблице: 0


#### Проверка столбцов с категориальными переменными

Обратимся к функции **register**, чтобы посмотреть сумму уникальных значений столбца **children**.

In [20]:
register('children') # посмотрим

 0     14091
 1      4808
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Выведем первые 10 строк таблицы, в которой значение столбца **children** соответствует **20**.

In [21]:
df[df['children'] == 20].head(10) # посмотрим

children  days_employed  dob_years education  education_id  \
606         20            880         21   среднее             1   
720         20            855         44   среднее             1   
1074        20           3310         56   среднее             1   
2510        20           2714         59    высшее             0   
2940        20           2161          0   среднее             1   
3300        20              0         35   среднее             1   
3394        20              0         56    высшее             0   
3669        20            913         23   среднее             1   
3695        20           2907         40   среднее             1   
3733        20            805         26    высшее             0   

              family_status  family_status_id gender  income_type  debt  \
606         женат / замужем                 0      M    компаньон     0   
720         женат / замужем                 0      F    компаньон     0   
1074        женат / замужем                 0      F    сотрудник     1   
2510         вдовец / вдова                 2      F    сотрудник     0   
2940        женат / замужем                 0      F    сотрудник     0   
3300  не женат / не замужем                 4      F  госслужащий     0   
3394        женат / замужем                 0      F    компаньон     0   
3669  не женат / не замужем                 4      F    сотрудник     0   
3695       гражданский брак                 1      M    сотрудник     0   
3733  не женат / не замужем                 4      M    сотрудник     0   

      total_income                                purpose  
606         145334                          покупка жилья  
720         112998                   покупка недвижимости  
1074        229518                  получение образования  
2510        264474  операции с коммерческой недвижимостью  
2940        199739                  на покупку автомобиля  
3300        136652                 профильное образование  
3394        201785                     высшее образование  
3669        101255     на покупку подержанного автомобиля  
3695        115380     на покупку подержанного автомобиля  
3733        137200                           ремонт жилью

Выведем первые 10 строк таблицы, в которой значение столбца **children** соответствует **-1**.

In [22]:
df[df['children'] == -1].head(20) # посмотрим

children  days_employed  dob_years education  education_id  \
291         -1           4417         46   среднее             1   
705         -1            902         50   среднее             1   
742         -1           3174         57   среднее             1   
800         -1         349987         54   среднее             1   
941         -1              0         57   среднее             1   
1363        -1           1195         55   среднее             1   
1929        -1           1461         38   среднее             1   
2073        -1           2539         42   среднее             1   
3812        -1           3045         26   среднее             1   
4198        -1            901         41   среднее             1   
4399        -1         398001         64   среднее             1   
4539        -1           1811         32   среднее             1   
5268        -1           1143         46   среднее             1   
6008        -1           1361         46    высшее             0   
6379        -1         370215         48   среднее             1   
7184        -1           5928         34   среднее             1   
7272        -1            526         51   среднее             1   
7609        -1              0         35   среднее             1   
7679        -1           3237         53   среднее             1   
8220        -1           1803         54    высшее             0   

              family_status  family_status_id gender  income_type  debt  \
291        гражданский брак                 1      F    сотрудник     0   
705         женат / замужем                 0      F  госслужащий     0   
742         женат / замужем                 0      F    сотрудник     0   
800   не женат / не замужем                 4      F    пенсионер     0   
941         женат / замужем                 0      F    пенсионер     0   
1363        женат / замужем                 0      F    компаньон     0   
1929  не женат / не замужем                 4      M    сотрудник     0   
2073              в разводе                 3      F    компаньон     0   
3812       гражданский брак                 1      F  госслужащий     0   
4198        женат / замужем                 0      F  госслужащий     0   
4399        женат / замужем                 0      F    пенсионер     0   
4539        женат / замужем                 0      F    сотрудник     0   
5268         вдовец / вдова                 2      F    сотрудник     0   
6008        женат / замужем                 0      F    сотрудник     0   
6379         вдовец / вдова                 2      F    пенсионер     0   
7184        женат / замужем                 0      M    сотрудник     0   
7272        женат / замужем                 0      F    сотрудник     0   
7609        женат / замужем                 0      M    сотрудник     0   
7679         вдовец / вдова                 2      F    сотрудник     0   
8220        женат / замужем                 0      F    компаньон     0   

      total_income                                 purpose  
291         102816                  профильное образование  
705         137882                 приобретение автомобиля  
742          64268              дополнительное образование  
800          86293              дополнительное образование  
941         114842            на покупку своего автомобиля  
1363         69550                  профильное образование  
1929        109121                           покупка жилья  
2073        162638                           покупка жилья  
3812        131892                   на проведение свадьбы  
4198        226375         операции со своей недвижимостью  
4399        163264                    покупка недвижимости  
4539        160544                             образование  
5268        278708                    покупка недвижимости  
6008        143008  строительство собственной недвижимости  
6379         36052                             образование  
7184        184315       

Из-за отсутствия закономерности, значения **20** и **-1** в столбце **children** заменим на **2** и **1** соответственно.

In [23]:
df.loc[df['children'] == 20,'children'] = 2 # 
df.loc[df['children'] == -1,'children'] = 1 # заменим 

Проверим значения столбца **dob_years**.

In [24]:
register('dob_years') # посмотрим

35    616
40    607
41    605
34    601
38    597
42    596
33    581
39    572
31    559
36    554
44    545
29    544
30    537
48    536
37    536
50    513
43    512
32    509
49    508
28    503
45    496
27    493
52    484
56    483
47    477
54    476
46    472
53    459
57    456
58    454
51    446
55    443
59    443
26    408
60    374
25    357
61    354
62    348
63    269
24    264
64    260
23    252
65    193
22    183
66    182
67    167
21    111
0     101
68     99
69     85
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Выведем первые 10 строк таблицы, в которой значение столбца **dob_years** соответствует **0**.

In [25]:
df[df['dob_years'] == 0].head(20) # посмотрим

children  days_employed  dob_years education  education_id  \
99           0         346541          0   среднее             1   
149          0           2664          0   среднее             1   
270          3           1872          0   среднее             1   
578          0         397856          0   среднее             1   
1040         0           1158          0    высшее             0   
1149         0            934          0   среднее             1   
1175         0         370879          0   среднее             1   
1386         0           5043          0    высшее             0   
1890         0              0          0    высшее             0   
1898         0         370144          0   среднее             1   
2082         0          10689          0   среднее             1   
2284         0              0          0   среднее             1   
2469         0            794          0   среднее             1   
2487         0           2497          0   среднее             1   
2869         0           1875          0   среднее             1   
2940         2           2161          0   среднее             1   
3043         2           2502          0    высшее             0   
3146         2            906          0    высшее             0   
3702         0           4288          0   среднее             1   
4062         1              0          0   среднее             1   

              family_status  family_status_id gender  income_type  debt  \
99          женат / замужем                 0      F    пенсионер     0   
149               в разводе                 3      F    сотрудник     0   
270         женат / замужем                 0      F    сотрудник     0   
578         женат / замужем                 0      F    пенсионер     0   
1040              в разводе                 3      F    компаньон     0   
1149        женат / замужем                 0      F    компаньон     0   
1175        женат / замужем                 0      F    пенсионер     0   
1386        женат / замужем                 0      M  госслужащий     0   
1890  не женат / не замужем                 4      F    сотрудник     0   
1898         вдовец / вдова                 2      F    пенсионер     0   
2082        женат / замужем                 0      F    сотрудник     0   
2284         вдовец / вдова                 2      F    пенсионер     0   
2469       гражданский брак                 1      F    сотрудник     0   
2487        женат / замужем                 0      F    сотрудник     1   
2869        женат / замужем                 0      F    компаньон     0   
2940        женат / замужем                 0      F    сотрудник     0   
3043  не женат / не замужем                 4      M    сотрудник     0   
3146        женат / замужем                 0      F    компаньон     0   
3702              в разводе                 3      F    сотрудник     0   
4062       гражданский брак                 1      M    компаньон     0   

      total_income                                 purpose  
99           71291                              автомобиль  
149          70176                       операции с жильем  
270         102166                            ремонт жилью  
578          97620  строительство собственной недвижимости  
1040        303994                         свой автомобиль  
1149        201852                    покупка недвижимости  
1175        313949   получение дополнительного образования  
1386        240523                    сделка с автомобилем  
1890        165640                                   жилье  
1898        127400                   на покупку автомобиля  
2082        386373                операции с недвижимостью  
2284        114842                            недвижимость  
2469         95967                         сыграть свадьбу  
2487         91367                            недвижимость  
2869        167713            на покупку своего автомобиля  
2940        199739       

В столбце **dob_years** отсутствует значение **18**. Ошибочное значение **0** маловероятно является **18**, так как значение **19** повторяется в таблице **14** раз, **20** - **51** раз, **21** - **111** раз, т.е. прогрессия. Из-за отсутствия закономерности ошибки заполнения в столбце **dob_years** значения **0** оставим без изменений.

Посмотрим значения в столбцах **education_id**, **family_status_id**, **debt** и **gender**.

In [26]:
display(register('education_id'))
display(register('family_status_id'))
display(register('debt'))
register('gender')# посмотрим

1    15172
0     5250
2      744
3      282
4        6
Name: education_id, dtype: int64

0    12339
1     4151
4     2810
3     1195
2      959
Name: family_status_id, dtype: int64

0    19713
1     1741
Name: debt, dtype: int64

F      14174
M       7279
XNA        1
Name: gender, dtype: int64

В столбцах **education_id**, **family_status_id** и **debt** отсутсвует некорректное заполнение.

В столбце **gender** одно значение `XNA`. Заменим его на **F**, так как доля **female** преобладает.

In [27]:
df.loc[df['gender'] == 'XNA','gender'] = 'F' # заменим

#### Лемматизация данных

In [28]:
# создадим переменную, в которую сохраним сумму уникальных значений столбца 'purpose'
credit_target = df['purpose'].value_counts().reset_index() 

credit_target # посмотрим

index  purpose
0                                  свадьба      791
1                    на проведение свадьбы      768
2                          сыграть свадьбу      765
3                 операции с недвижимостью      675
4        покупка коммерческой недвижимости      661
5                        операции с жильем      652
6                  покупка жилья для сдачи      651
7    операции с коммерческой недвижимостью      650
8                                    жилье      646
9                            покупка жилья      646
10                 покупка жилья для семьи      638
11  строительство собственной недвижимости      635
12                            недвижимость      633
13         операции со своей недвижимостью      627
14        строительство жилой недвижимости      624
15                    покупка недвижимости      621
16                    покупка своего жилья      620
17              строительство недвижимости      619
18                            ремонт жилью      607
19              покупка жилой недвижимости      606
20            на покупку своего автомобиля      505
21            заняться высшим образованием      496
22                              автомобиль      494
23        сделка с подержанным автомобилем      486
24                              автомобили      478
25                         свой автомобиль      478
26      на покупку подержанного автомобиля      478
27                   на покупку автомобиля      471
28                 приобретение автомобиля      461
29              дополнительное образование      460
30                    сделка с автомобилем      455
31                      высшее образование      452
32                             образование      447
33   получение дополнительного образования      446
34                   получение образования      442
35                  профильное образование      436
36           получение высшего образования      426
37                   заняться образованием      408

Создадим функцию **credit_lemmatize**, которая будет производить лемматизацию.

In [29]:
m = Mystem()

def credit_lemmatize(i):
    lemmas = m.lemmatize(i)
    return lemmas

credit_target['credit_lemmatize'] = credit_target['index'].apply(credit_lemmatize)

credit_target # посмотрим

index  purpose  \
0                                  свадьба      791   
1                    на проведение свадьбы      768   
2                          сыграть свадьбу      765   
3                 операции с недвижимостью      675   
4        покупка коммерческой недвижимости      661   
5                        операции с жильем      652   
6                  покупка жилья для сдачи      651   
7    операции с коммерческой недвижимостью      650   
8                                    жилье      646   
9                            покупка жилья      646   
10                 покупка жилья для семьи      638   
11  строительство собственной недвижимости      635   
12                            недвижимость      633   
13         операции со своей недвижимостью      627   
14        строительство жилой недвижимости      624   
15                    покупка недвижимости      621   
16                    покупка своего жилья      620   
17              строительство недвижимости      619   
18                            ремонт жилью      607   
19              покупка жилой недвижимости      606   
20            на покупку своего автомобиля      505   
21            заняться высшим образованием      496   
22                              автомобиль      494   
23        сделка с подержанным автомобилем      486   
24                              автомобили      478   
25                         свой автомобиль      478   
26      на покупку подержанного автомобиля      478   
27                   на покупку автомобиля      471   
28                 приобретение автомобиля      461   
29              дополнительное образование      460   
30                    сделка с автомобилем      455   
31                      высшее образование      452   
32                             образование      447   
33   получение дополнительного образования      446   
34                   получение образования      442   
35                  профильное образование      436   
36           получение высшего образования      426   
37                   заняться образованием      408   

                                     credit_lemmatize  
0                                       [свадьба, \n]  
1                 [на,  , проведение,  , свадьба, \n]  
2                           [сыграть,  , свадьба, \n]  
3               [операция,  , с,  , недвижимость, \n]  
4     [покупка,  , коммерческий,  , недвижимость, \n]  
5                      [операция,  , с,  , жилье, \n]  
6           [покупка,  , жилье,  , для,  , сдача, \n]  
7   [операция,  , с,  , коммерческий,  , недвижимо...  
8                                         [жилье, \n]  
9                             [покупка,  , жилье, \n]  
10          [покупка,  , жилье,  , для,  , семья, \n]  
11  [строительство,  , собственный,  , недвижимост...  
12                                 [недвижимость, \n]  
13    [операция,  , со,  , свой,  , недвижимость, \n]  
14     [строительство,  , жилой,  , недвижимость, \n]  
15                     [покупка,  , недвижимость, \n]  
16                   [покупка,  , свой,  , жилье, \n]  
17               [строительство,  , недвижимость, \n]  
18                             [ремонт,  , жилье, \n]  
19           [покупка,  , жилой,  , недвижимость, \n]  
20       [на,  , покупка,  , свой,  , автомобиль, \n]  
21       [заниматься,  , высокий,  , образование, \n]  
22                                   [автомобиль, \n]  
23  [сделка,  , с,  , подержанный,  , автомобиль, \n]  
24                                   [автомобиль, \n]  
25                          [свой,  , автомобиль, \n]  
26  [на,  , покупка,  , подержать,  , автомобиль, \n]  
27                [на,  , покупка,  , автомобиль, \n]  
28                  [приобретение,  , автомобиль, \n]  
29               [дополнительный,  , образование, \n]  
30                  [сделка,  , с,  , автомобиль, \n]  
31                      [высокий,  , образование, \n]  
32                                  [образо

По итогу лемматизации, резделим столбец **purpose** на **4** категорий:

1. Проведения свадьбы
2. Операции с недвижимостью
3. Операции с автомобилем
4. Получение образования

#### Категоризация столбца 'purpose'

Создадим функцию **credit_target** для замены значений.

In [30]:
def credit_target(purpose):
    if 'образован' in purpose:
        return 'образование'
    elif 'авто' in purpose:
        return 'автомобиль'
    elif 'свадь' in purpose:
        return 'свадьба'
    else:
        return 'недвижимость'
    

df['credit_target'] = df['purpose'].apply(credit_target) # создадим столбец 'credit_target'

df[['purpose', 'credit_target']].head(10) # проверим

purpose credit_target
0               покупка жилья  недвижимость
1     приобретение автомобиля    автомобиль
2               покупка жилья  недвижимость
3  дополнительное образование   образование
4             сыграть свадьбу       свадьба
5               покупка жилья  недвижимость
6           операции с жильем  недвижимость
7                 образование   образование
8       на проведение свадьбы       свадьба
9     покупка жилья для семьи  недвижимость

Создадим столбец **credit_target_id**, в котором присвоим номер категории для целей кредита.

In [31]:
def credit_target_id(credit_target):
    if 'образование' in credit_target:
        return 0
    elif 'автомобиль' in credit_target:
        return 1
    elif 'свадьба' in credit_target:
        return 2
    else:
        return 3

df['credit_target_id'] = df['credit_target'].apply(credit_target_id)

df[['credit_target', 'credit_target_id']].head(10) # проверим

credit_target  credit_target_id
0  недвижимость                 3
1    автомобиль                 1
2  недвижимость                 3
3   образование                 0
4       свадьба                 2
5  недвижимость                 3
6  недвижимость                 3
7   образование                 0
8       свадьба                 2
9  недвижимость                 3

In [32]:
df.drop('purpose', axis=1, inplace=True) # удалим первоначальный столбец 'purpose'

В результате мы изменили множественные пересекающиеся значения столбца **purpose** на **4** типа (столбец **credit_target**) и присвоили каждому типу категорию (столбец **credit_target_id**).

### Создание словарей

Выделим и сформируем три словаря **df_dict_education**, **df_dict_family**, **df_dict_credit_target_id**. В предстоящих исследованиях обращусь к ним для проверки некоторых гитопез.

In [33]:
# создадим ключи для словарей
df_keys_education = df[['education_id']] 
df_keys_family = df[['family_status_id']]
df_keys_credit_target_id = df[['credit_target_id']]

# создадим словари для 'education', 'family_status' и 'credit_target'
df_dict_education = df[['education_id', 'education']]
df_dict_family = df[['family_status_id', 'family_status']]
df_dict_credit_target_id = df[['credit_target_id', 'credit_target']]

# удалим дупликаты
df_dict_education = df_dict_education.drop_duplicates().reset_index(drop=True)
df_dict_family = df_dict_family.drop_duplicates().reset_index(drop=True) 
df_dict_credit_target_id = df_dict_credit_target_id.drop_duplicates().reset_index(drop=True) 

# выведем результат на экран
display(df_dict_education)
display(df_dict_family)
df_dict_credit_target_id

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

credit_target_id credit_target
0                 3  недвижимость
1                 1    автомобиль
2                 0   образование
3                 2       свадьба

## Проверка гипотез

Перед нами поставлена задача — проверить 4 гипотезы:

1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?

### Проверка зависимости между наличием детей и возвратом кредита в срок

In [34]:
df_children = df.pivot_table(
    index = 'children',
    values = 'debt',
    aggfunc = ['count', 'sum', 'mean']
)

df_children

count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  0.075438
1          4855   445  0.091658
2          2128   202  0.094925
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

In [35]:
(df_children['sum'].sum() - 1063) / (df_children['count'].sum() - 14091)

debt    0.092082
dtype: float64

In [36]:
print(f'Доля заемщиков с задолженностью без детей: {0.075438:.2%}') 
print(f'Доля заемщиков с задолженностью с детьми: {0.092082:.2%}')

Доля заемщиков с задолженностью без детей: 7.54%
Доля заемщиков с задолженностью с детьми: 9.21%


**Вывод:**

- процент заемщиков с задолженностью с детьми больше, чем процент заемщиков с задолженностью без детей (**9.21%** против **7.54%**).

### Проверка зависимости между семейным положением и возвратом кредита в срок

In [37]:
df.pivot_table(
    index = 'family_status',
    values = 'debt',
    aggfunc = ['count', 'sum', 'mean']
)

count  sum      mean
                        debt debt      debt
family_status                              
в разводе               1195   85  0.071130
вдовец / вдова           959   63  0.065693
гражданский брак        4151  388  0.093471
женат / замужем        12339  931  0.075452
не женат / не замужем   2810  274  0.097509

**Вывод:**
- гипотеза подтвердилась: есть зависимость между семейным положением и возвратом кредита в срок;
- наше исследование показало, что самая большая доля заемщиков имеющие долги по кредитам - это **не женат / не замужем** и **гражданский брак** (**9.75%** и **9.35%** соответственно);
- средние показатели, в рамках данного исследования, имеют заемщики **женат / замужем** и **в разводе** (**7.55%** и **7.11%** соответственно);
- результат нашего исследования показал, что самые надежные заемщики - это люди имеющие в **family_status** **вдовец / вдова** и доля задолженности по кредитам составляет всего **6.57%**.

### Проверка зависимости между уровнем дохода и возратом кредита в срок

Создадим столбец, который присваивает категорию в зависимости от значения в столбце **total_income**.

In [38]:
def income(a):
    if a < 100000:
        return 0
    elif 100000 < a < 150000:
        return 1
    elif 150000 < a < 200000:
        return 2
    else:
        return 3

df['total_income_id'] = df['total_income'].apply(income) # создадим столбец

In [39]:
df.pivot_table(
    index = 'total_income_id',
    values = 'debt',
    aggfunc = ['count', 'sum', 'mean']
)

count  sum      mean
                 debt debt      debt
total_income_id                     
0                4463  354  0.079319
1                6971  616  0.088366
2                4767  405  0.084959
3                5253  366  0.069674

**Вывод:**
- гипотеза подтвердилась: есть зависимость между уровнем дохода и возвратом кредита в срок;
- наше исследование показало, что самая большая доля заемщиков имеющие долги по кредитам - это люди с доходом в промежутке **между 100000 и 200000** (**8.84%** и **8.5%** соответственно);
- чуть меньше доля задолженности у заемщиков с доходом **до 100000** (**7.93%**);
- результат нашего исследования показал, что самые надежные заемщики - это люди имеющие доход **более 200000** и доля задолженности по кредитам составляет **6.97%**.

### Проверка завимости целей кредита влияют на возврат в срок

In [40]:
df.pivot_table(
    index = 'credit_target',
    values = 'debt',
    aggfunc = ['count', 'sum', 'mean']
)

count  sum      mean
                debt debt      debt
credit_target                      
автомобиль      4306  403  0.093590
недвижимость   10811  782  0.072334
образование     4013  370  0.092200
свадьба         2324  186  0.080034

**Вывод:**

- гипотеза подтвердилась: есть зависимость между разными целями и возвратом кредита в срок;
- наше исследование показало, что самая большая доля заемщиков имеющие долги по кредитам - это люди с целью кредита **образование** и **автомобиль** (**9.22%** и **9.33%** соответственно);
- средние показатели, в рамках данного исследования, имеют заемщики с целью кредита **свадьба** (**8%**);
- результат нашего исследования показал, что самые надежные заемщики - это люди у которых цель кредита **недвижимость** и доля задолженности по кредитам составляет всего **7.23%**.

## Итоги исследования

Мы проверили четыре гипотезы и установили:

1. Есть ли зависимость между наличием детей и возвратом кредита в срок?
Первая гипотеза была опровергнута. Результаты исследований показали, что существует обратная зависимость, а именно: заемщики с детьми чаще оказываются должниками нежели заемщики без детей.

2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
Гипотеза полностью подтвердилась! Самая большая доля заемщиков с долгом по кредиту в статусе: **не женат / не замужем** и **гражданский брак**. Средние показатели у **женат / замужем** и **в разводе** и самые надежные заемщики имеют статус **вдовец / вдова**.

3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
Гипотеза полностью подтвердилась! Самая большая доля заемщиков с долгом по кредиту имеет доход **от 100000 до 200000**. Средние показатели у с доходом **до 100000** и самые надежные заемщики имеют доход **более 200000**.

4. Как разные цели кредита влияют на его возврат в срок?
Самая большая доля заемщиков с долгом по кредиту брали кредит на: **образование** и **автомобиль**. Средние показатели у цели на **свадьбу** и самые надежные заемщики берут кредит на **недвижимость**.

По итогу исследования мы можем составить группы заемщиков:

1. В группе **семейное положение**: самые ответственные заемщики имеют статус **вдовец / вдова** и **6.57%** невозврата, а самые частые должники - **не женат / не замужем** и **9.75%** невозврата соответственно.
2. В группе **доход**: самые ответственные заемщики имеют доход **более 200000** и **6.97%** невозврата, а самые частые должники - это люди с доходом **от 100000 до 150000** и **8.84%** невозврата соответственно.
3. В группе **цель кредита**: самые ответственные заемщики берут кредит на **недвижимость** и **7.23%** невозврата, а самые частые должники с целью **автомобиль** и **9.33%** невозврата соответственно.
